lib/restforce/db/associator.rb

```diff
@@ -0,0 +1,104 @@
+        @mapping = mapping
+        @runner = runner
+      end
+
+      # Public: Run the re-association process, pulling in records from
+      # Salesforce and the database to determine the most recently attached
+      # association, then propagating the change between systems.
+      #
+      # Returns nothing.
+      def run
+        return if associations.empty?
+
+        @runner.run(@mapping) do |run|
+          run.salesforce_instances { |instance| verify_associations(instance) }
+          run.database_instances { |instance| verify_associations(instance) }
+        end
+      end
+
+      private
+
+      # Internal: Ensure integrity between the lookup columns in Salesforce and
+      # the synchronized records in the database.
+      #
+      # instance - A Restforce::DB::Instances::Base.
+      #
+      # Returns nothing.
+      def verify_associations(instance)
+        database_instance, salesforce_instance =
+          case instance
+          when Restforce::DB::Instances::Salesforce
+            [@mapping.database_record_type.find(instance.id), instance]
+          when Restforce::DB::Instances::ActiveRecord
+            [instance, @mapping.salesforce_record_type.find(instance.id)]
+          end
+
+        return unless database_instance && salesforce_instance
+        sync_associations(database_instance, salesforce_instance)
+      end
+
+      # Internal: Given a database record and corresponding Salesforce data,
+      # synchronize the record associations in whichever system has the least
+      # recent data.
+      #
+      # database_instance   - A Restforce::DB::Instances::ActiveRecord.
+      # salesforce_instance - A Restforce::DB::Instances::Salesforce.
+      #
+      # Returns nothing.
+      def sync_associations(database_instance, salesforce_instance)
+        ids = association_ids(database_instance)
+        return if ids.all? { |field, id| salesforce_instance.record[field] == id }
+
+        if database_instance.last_update > salesforce_instance.last_update
+          salesforce_instance.update!(ids)
+        else
+          database_record = database_instance.record
+          associations.each do |association|
+            association.build(database_record, salesforce_instance.record)
+          end
+          database_record.save!
+        end
+      end
+
+      # Internal: Get a Hash of associated lookup IDs for the passed database
+      # record.
+      #
+      # database_instance - A Restforce::DB::Instances::ActiveRecord.
+      #
+      # Returns a Hash.
+      def association_ids(database_instance)
+        associations.inject({}) do |ids, association|
+          ids.merge(association.lookups(database_instance.record))
+        end
+      end
+
+      # Internal: Get a list of the BelongsTo associations defined for the
+      # target mapping.
+      #
+      # Returns an Array of Restforce::DB::Association::BelongsTo objects.
+      def associations
```

Comment:
Possibly. That would be more explicit, but I named it with the intent of "this is all the associations we actually care about". The fact that we only currently care about `belongs_to` associations is sort of auxiliary. I can change it if you feel strongly.